gpu_bdb/queries/q04/gpu_bdb_query_04_dask_sql.py

```diff
@@ -0,0 +1,176 @@
+        .str.cat(filtered_df["wp_type_codes"].astype("str"), sep=":")
+    )
+    # This gives the last occurrence (by timestamp) within the "order", "dynamic" wp_types
+    filtered_df = filtered_df.groupby(
+        ["wcs_user_sk", "session_id"], as_index=False, sort=False
+    ).agg({"wp_type_codes": "max"})
+    # If the max contains dynamic, keep the row else discard.
+    last_dynamic_df = filtered_df[
+        filtered_df["wp_type_codes"].str.contains(
+            ":" + str(DYNAMIC_CAT_CODE), regex=False
+        )
+    ]
+    del filtered_df
+
+    # Find counts for each group
+    grouped_count_df = df.groupby(
+        ["wcs_user_sk", "session_id"], as_index=False, sort=False
+    ).agg({"tstamp_inSec": "count"})
+    # Merge counts with the "dynamic" shopping cart groups
+    result = last_dynamic_df.merge(
+        grouped_count_df, on=["wcs_user_sk", "session_id"], how="inner"
+    )
+    del (last_dynamic_df, grouped_count_df)
+    return cudf.DataFrame(
+        {"pagecount": result.tstamp_inSec.sum(), "count": len(result)}
+    )
+
+
+def reduction_function(df, keep_cols, DYNAMIC_CAT_CODE, ORDER_CAT_CODE):
+    df = get_sessions(df, keep_cols=keep_cols)
+    df = abandonedShoppingCarts(
+        df, DYNAMIC_CAT_CODE=DYNAMIC_CAT_CODE, ORDER_CAT_CODE=ORDER_CAT_CODE
+    )
+    return df
+
+
+def read_tables(data_dir, c, config):
+    table_reader = build_reader(
+        data_format=config["file_format"],
+        basepath=config["data_dir"],
+        split_row_groups=config["split_row_groups"],
+    )
+
+    wp_cols = ["wp_type", "wp_web_page_sk"]
+    wp_df = table_reader.read("web_page", relevant_cols=wp_cols)
+
+    wcs_cols = [
+        "wcs_user_sk",
+        "wcs_click_date_sk",
+        "wcs_click_time_sk",
+        "wcs_web_page_sk",
+        "wcs_sales_sk",
+    ]
+    wcs_df = table_reader.read("web_clickstreams", relevant_cols=wcs_cols)
+
+    c.create_table('web_page_wo_categorical', wp_df, persist=False)
+    c.create_table('web_clickstreams', wcs_df, persist=False)
+
+
+def main(data_dir, client, c, config):
+    benchmark(read_tables, data_dir, c, config, dask_profile=config["dask_profile"])
+
+    query_web_page = """
+        SELECT wp_type, wp_web_page_sk
+        FROM web_page_wo_categorical
+    """
+    wp = c.sql(query_web_page)
+
+    # Convert wp_type to categorical and get cat_id of review and dynamic type
+    wp["wp_type"] = wp["wp_type"].map_partitions(
+                                    lambda ser: ser.astype("category"))
+    
+    cpu_categories = wp["wp_type"].compute().cat.categories
+    if hasattr(cpu_categories, "to_pandas"):
+        cpu_categories = cpu_categories.to_pandas()
+
+    DYNAMIC_CAT_CODE = cpu_categories.get_loc("dynamic")
+    ORDER_CAT_CODE = cpu_categories.get_loc("order")
+
+    # ### cast to minimum viable dtype
+    import cudf
+    codes_min_signed_type = cudf.utils.dtypes.min_signed_type(
+                                                    len(cpu_categories))
```

Comment:
No longer needed.

Reply:
We don't need the typecasts to `codes_min_signed_type`?

Reply:
Yup.  cuDF now does that  on its own. 

```python

import cudf
gdf = cudf.DataFrame({"id": [1, 2, 3, 4, 5, 6], "grade":['a', 'b', 'b', 'a', 'a', 'e']})
gdf['grade'] = gdf['grade'].astype('category')
print(type(gdf['grade'].cat.codes.dtype))
```
```
<class 'numpy.dtype[uint8]'>
```

Reply:
Added a comment below . Resolving this conversation.